presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/CommitCleaner.java

```diff
@@ -0,0 +1,189 @@
+
+        this.masterDao = createJdbi(database.getMasterConnection())
+                .onDemand(MasterCommitCleanerDao.class);
+
+        this.deletedChunksDao = createDeletedChunksDao(database);
+
+        this.shardDbi = database.getShards().stream()
+                .map(shard -> createJdbi(shard.getConnection()))
+                .collect(toImmutableList());
+    }
+
+    public synchronized void removeOldCommits()
+    {
+        long activeCommitId = transactionManager.oldestActiveCommitId();
+
+        Set<Long> droppedTableIds = masterDao.getDroppedTableIds(activeCommitId);
+
+        Map<Integer, Set<Long>> droppedTableIdsByShard = droppedTableIds.stream()
+                .collect(groupingBy(tableId -> tableShard(tableId, shardDbi.size()), toSet()));
+
+        // drop index tables
+        for (long tableId : droppedTableIds) {
+            try (Handle handle = shardDbi.get(tableShard(tableId, shardDbi.size())).open()) {
+                handle.execute("DROP TABLE IF EXISTS " + chunkIndexTable(tableId));
+            }
+        }
+
+        // cleanup deleted chunks
+        long now = System.currentTimeMillis();
+        for (int i = 0; i < shardDbi.size(); i++) {
+            ShardCommitCleanerDao shard = shardDbi.get(i).onDemand(ShardCommitCleanerDao.class);
+            Set<Long> tableIds = droppedTableIdsByShard.getOrDefault(i, ImmutableSet.of());
+
+            // cleanup chunks for old commits
+            List<TableChunk> chunks;
+            do {
+                chunks = shard.getDeletedChunks(activeCommitId, CHUNK_BATCH_SIZE);
+                if (!chunks.isEmpty()) {
+                    cleanupChunks(shard, now, chunks, tableIds);
+                }
+            }
+            while (chunks.size() == CHUNK_BATCH_SIZE);
+
+            // cleanup chunks for dropped tables
+            if (!tableIds.isEmpty()) {
+                do {
+                    chunks = shard.getDeletedChunks(tableIds, CHUNK_BATCH_SIZE);
+                    if (!chunks.isEmpty()) {
+                        cleanupChunks(shard, now, chunks, tableIds);
+                    }
+                }
+                while (chunks.size() == CHUNK_BATCH_SIZE);
+            }
+        }
+
+        // cleanup dropped table columns
+        if (!droppedTableIds.isEmpty()) {
+            masterDao.cleanupDroppedTableColumns(droppedTableIds);
+        }
+
+        // cleanup schemas, tables, columns, views, commits
+        masterDao.cleanup(activeCommitId);
+
+        // cleanup table sizes
+        for (Jdbi shard : shardDbi) {
+            shard.onDemand(ShardCommitCleanerDao.class)
+                    .cleanupTableSizes(activeCommitId);
+        }
+    }
+
+    private void cleanupChunks(ShardCommitCleanerDao shard, long now, List<TableChunk> chunks, Set<Long> droppedTableIds)
+    {
+        // queue for deletion
+        deletedChunksDao.insertDeletedChunks(now, chunks);
```

Comment:
An unsafe state may happen after lien 129 and before line 139.

After line 129, the delete_chunks table has content, and ChunkStoreCleaner will remove data. So the unsafe state is 
some metadata is still there,   but data is not.

Although this may not cause any issue for now, usually we remove all such states in storage system. If we continue adding codes, we may trigger more unsafe states that cause actual issue.

In v1, we actually did some delay to prevent it, if you check
shardCleaner.java    
private static Timestamp maxTimestamp(Duration duration)

I think we should use the same v1 logic.

Reply:
This is a good observation, but the code is safe. None of the cleanup work here is atomic, but it is idempotent. We expect it can fail at any point and cleanup will continue on the next run. The data being cleaned up here is not observable by anyone.

The delay in v1 has a more important reason: we don't track active transactions, so we want to avoid deleting something that an active query might still need to read.

Reply:
Note that in either v1 or v2, we never use timestamps for correctness. We always assume that clocks could jump backwards or forwards randomly. Timestamps are only used for liveness. In v1, if the clock jumps forward, we might delete data that a query needs, but that query will explicitly fail -- what wouldn't be ok is to skip the data silently.

Reply:
The same is true for cleaning up local data on workers. There is a race between when the file is written locally and when it is committed into the metadata. We make a tradeoff between how fast we want to delete data and how often we will need to recover when we make a mistake (due to insert queries taking a very long time or clock anomalies).

Reply:
Now I understand why we had it in V1. 

I believe this is valid too:  "The data being cleaned up here is not observable by anyone"